src/zenml/integrations/seldon/services/seldon_deployment.py

```diff
@@ -42,45 +44,107 @@ class SeldonDeploymentConfig(ServiceConfig):
-    #   folders and sub-folders depending on the model type/format and the
-    #   Seldon Core protocol used to serve the model.
-    model_uri: str
-    model_name: str
-    # TODO [ENG-774]: have an enum of all model formats ?
-    model_format: Optional[str]
-    # TODO [ENG-775]: have an enum of all supported Seldon Core implementation ?
+    model_uri: str = ""
+    model_name: str = "default"
+    # TODO [ENG-775]: have an enum of all supported Seldon Core implementations
     implementation: str
-    pipeline_name: Optional[str] = None
-    pipeline_run_id: Optional[str] = None
-    pipeline_step_name: Optional[str] = None
     replicas: int = 1
+    secret_name: Optional[str]
     model_metadata: Dict[str, Any] = Field(default_factory=dict)
     extra_args: Dict[str, Any] = Field(default_factory=dict)
 
-    # configuration attributes that are not part of the service configuration
-    # but are required for the service to function. These must be moved to the
-    # stack component, when available
-    kubernetes_context: Optional[str]
-    namespace: Optional[str]
-    base_url: str
-    # TODO [ENG-776]: replace with ZenML secret and create a k8s secret resource
-    #   that can be mounted in the container
-    secret_name: Optional[str]
+    def get_seldon_deployment_labels(self) -> Dict[str, str]:
+        """Generate the labels for the Seldon Core deployment from the
+        service configuration.
+
+        These labels are attached to the Seldon Core deployment resource
+        and may be used as label selectors in lookup operations.
+
+        Returns:
+            The labels for the Seldon Core deployment.
+        """
+        labels = {}
+        if self.pipeline_name:
+            labels["zenml.pipeline_name"] = self.pipeline_name
+        if self.pipeline_run_id:
+            labels["zenml.pipeline_run_id"] = self.pipeline_run_id
+        if self.pipeline_step_name:
+            labels["zenml.pipeline_step_name"] = self.pipeline_step_name
+        if self.model_name:
+            labels["zenml.model_name"] = self.model_name
+        if self.model_uri:
+            labels["zenml.model_uri"] = self.model_uri
+        if self.implementation:
+            labels["zenml.model_type"] = self.implementation
+        SeldonClient.sanitize_labels(labels)
+        return labels
+
+    def get_seldon_deployment_annotations(self) -> Dict[str, str]:
+        """Generate the annotations for the Seldon Core deployment from the
+        service configuration.
+
+        The annotations are used to store additional information about the
+        Seldon Core service that is associated with the deployment that is
+        not available in the labels. One annotation particularly important
+        is the serialized Service configuration itself, which is used to
+        recreate the service configuration from a remote Seldon deployment.
+
+        Returns:
+            The annotations for the Seldon Core deployment.
```

Comment:
Is this annotation vs. labels difference a seldon core concept?

Reply:
Labels and annotations are Kubernetes concepts. They can be "attached" to any Kubernetes resource, including the SeldonDeployment.

Labels are used as simple key-value tags that can be used to perform advanced searches like listing all the pods that belong to a particular application (i.e. including the `app` label as a label selector in your `kubectl get` command).

Annotations can be used to store more complex information in your Kubernetes resource. (think of them as metadata). We use an annotation to store the entire Service configuration serialized inside the SeldonDeployment resource. We can re-create the Service from the SeldonDeployment resource this way when we fetch the resources from the kube cluster.

Reply:
here is an example:

```yaml
  metadata:
    annotations:
      zenml.service_config: '{"type": "zenml.integrations.seldon.services.seldon_deployment.SeldonDeploymentConfig",
        "name": "", "description": "", "pipeline_name": "continuous_deployment_pipeline",
        "pipeline_run_id": "continuous_deployment_pipeline-08_Apr_22-13_04_10_875604",
        "pipeline_step_name": "seldon_model_deployer_step", "model_uri": "s3://zenfiles/seldon_model_deployer_step/output/854/seldon",
        "model_name": "mnist", "implementation": "SKLEARN_SERVER", "replicas": 1,
        "secret_name": "seldon-init-container-secret", "model_metadata": {}, "extra_args":
        {}}'
      zenml.version: 0.7.0
    creationTimestamp: "2022-04-07T20:55:58Z"
    generation: 5
    labels:
      app: zenml
      zenml.model_name: mnist
      zenml.model_type: SKLEARN_SERVER
      zenml.model_uri: s3_zenfiles_seldon_model_deployer_step_output_854_seldon
      zenml.pipeline_name: continuous_deployment_pipeline
      zenml.pipeline_run_id: continuous_deployment_pipeline-08_Apr_22-13_04_10_875604
      zenml.pipeline_step_name: seldon_model_deployer_step
      zenml.service_uuid: 22c6c8ed-bccd-45fc-80a5-d7e2143d7e09
    name: zenml-22c6c8ed-bccd-45fc-80a5-d7e2143d7e09
```

Reply:
Thanks for the detailed answer, makes sense :)